WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift

```diff
@@ -0,0 +1,109 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireAPI
+import WireDataModel
+
+/// An extension that encapsulates storage operations related to conversation MLS.
+
+extension ConversationLocalStore {
+
+    // MARK: - Message protocols
+
+    func assignMessageProtocol(
+        from remoteConversation: WireAPI.Conversation,
+        for localConversation: ZMConversation
+    ) {
+        guard let newMessageProtocol = remoteConversation.messageProtocol else {
+            return
+        }
+
+        localConversation.messageProtocol = newMessageProtocol.toDomainModel()
+    }
+
+    func updateMessageProtocol(
+        from remoteConversation: WireAPI.Conversation,
+        for localConversation: ZMConversation
+    ) {
+        guard let newMessageProtocol = remoteConversation.messageProtocol else {
+            return
+        }
+
+        let sender = ZMUser.selfUser(in: context)
+
+        switch localConversation.messageProtocol {
+        case .proteus:
+            switch newMessageProtocol {
+            case .proteus:
+                break /// no update, ignore
+            case .mixed:
+                localConversation.appendMLSMigrationStartedSystemMessage(sender: sender, at: .now)
+                localConversation.messageProtocol = newMessageProtocol.toDomainModel()
+
+            case .mls:
+                let date = localConversation.lastModifiedDate ?? .now
+                localConversation.appendMLSMigrationPotentialGapSystemMessage(sender: sender, at: date)
+                localConversation.messageProtocol = newMessageProtocol.toDomainModel()
+            }
+
+        case .mixed:
+            switch newMessageProtocol {
+            case .proteus, .mixed:
+                break /// no update, ignore
+            case .mls:
+                localConversation.appendMLSMigrationFinalizedSystemMessage(sender: sender, at: .now)
+                localConversation.messageProtocol = newMessageProtocol.toDomainModel()
+            }
+
+        case .mls:
+            switch newMessageProtocol {
+            case .proteus, .mixed, .mls:
+                break
+            }
+        }
+    }
+
+    // MARK: - Self / MLS
+
+    func createOrJoinSelfConversation(
+        from localConversation: ZMConversation
+    ) async throws {
+        guard let context = localConversation.managedObjectContext else {
```

Comment:
Can you not use the context that belongs to self here? That way it can't be `nil`.

Reply:
will do thx

Reply:
fixed